File: JAVA/948. Bag of Tokens.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {        
        int n = tokens.length;
        int maxScore = 0;
        Arrays.sort(tokens);
        int i = 0, j = n-1;
        int score = 0;
        while(i <= j) {
            if(power >= tokens[i]) {
                power -= tokens[i];
                score += 1;
                i++;
                maxScore = Math.max(maxScore, score);
            }
            else if(score >= 1) {
                power += tokens[j];
                score -= 1;
                j--;
            }
            else {
                return maxScore;
            }
        } 
        return maxScore;
    }
}

File: C++/948. Bag of Tokens.cpp
class Solution {
public:
    int bagOfTokensScore(std::vector<int>& tokens, int power) {
        std::sort(tokens.begin(), tokens.end());        
        int score = 0, maxScore = 0;
        int left = 0, right = tokens.size() - 1;        
        while (left <= right) {
            if (power >= tokens[left]) {
                power -= tokens[left++];
                score++;
                maxScore = std::max(maxScore, score);
            }
            else if (score > 0) {
                power += tokens[right--];
                score--;
            }
            else {
                break;
            }
        }       
        return maxScore;
    }
};
